Add e2e backup and restore test suite for oadp-cli

Joeavaikath opened the pull request:
## Why the changes were made

Adding e2e test suite for the oadp-cli. 

## How to test the changes made

Can run the "Backup and restore tests via OADP CLI" ginkgo test suite.

Joeavaikath committed:
start cli e2e work

Joeavaikath committed:
First pass done

kaovilai commented:
/test-all

kaovilai commented:
/test all

Joeavaikath committed:
Install oadp-cli in CI image

Joeavaikath commented:
/test-all

Joeavaikath commented:
/test all

Joeavaikath committed:
Fix linting

Joeavaikath commented:
/retest

Joeavaikath committed:
Refactor code

kaovilai commented on tests/e2e/lib/backup_cli.go:
```diff
@@ -0,0 +1,261 @@
+// CreateBackupForNamespacesViaCLI creates a backup using the OADP CLI
+func CreateBackupForNamespacesViaCLI(backupName string, namespaces []string, defaultVolumesToFsBackup bool, snapshotMoveData bool) error {
+	var options []string
+
+	// Add included namespaces (comma-separated)
+	if len(namespaces) > 0 {
+		options = append(options, "--include-namespaces", strings.Join(namespaces, ","))
```
This func sounded like it would already contain `oadp` in the exec.

Joeavaikath replied:
oadp cli is installed on the docker image, so it is?

kaovilai replied:
I guess I was commenting on some older commits.. resolving.

Joeavaikath commented:
/retest

kaovilai commented on tests/e2e/backup_restore_cli_suite_test.go:
```diff
@@ -0,0 +1,377 @@
+		cmd := exec.Command("kubectl", "oadp", "version")
+		output, err := cmd.CombinedOutput()
+		if err != nil {
+			ginkgo.Skip(fmt.Sprintf("OADP CLI not available: %v, output: %s", err, string(output)))
```
You might have to install also kubectl

```
  [SKIPPED] OADP CLI not available: exec: "kubectl": executable file not found in $PATH, output: 
```
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_oadp-operator/1835/pull-ci-openshift-oadp-operator-master-4.19-e2e-test-aws/1948023401226440704/artifacts/e2e-test-aws/e2e/build-log.txt#:~:text=%5BSKIPPED%5D%20OADP%20CLI%20not%20available%3A%20exec%3A%20%22kubectl%22%3A%20executable%20file%20not%20found%20in%20%24PATH%2C%20output%3A

Joeavaikath replied:
Ah yes, will do

kaovilai replied:
Perhaps ginkgo.Fail if this is in openshift ci env but that can come later.. but def false postive atm.

Joeavaikath replied:
Pushed the ci image with kubectl lets see

Joeavaikath committed:
Add kubectl

kaovilai commented:
whoops test timed out..

kaovilai commented:
In case you are interested, timeout code is [here](https://github.com/kubernetes-sigs/prow/blob/d01b8af18f00474804250403f9bc3d8e45852190/pkg/entrypoint/run.go#L189). Will probably need to bump up timeouts if any configured for this repo. Alternatively, instead of adding on tests, we might consider replacing some of the existing tests.

kaovilai commented:
https://github.com/openshift/release/pull/67421

Joeavaikath commented:
/retest

Joeavaikath committed:
Add TEST_CLI option to Makefile and update testing documentation

- Introduced TEST_CLI variable to control CLI-based backup/restore testing in the Makefile.
- Updated TESTING.md to include information about the new TEST_CLI environment variable.
- Added 'cli' label to the CLI test suite for better categorization.

Joeavaikath commented:
/retest

kaovilai reviewed (changes requested):
```
Backup and restore tests via OADP CLI [cli]
/go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_cli_suite_test.go:188
  Backup and restore applications via OADP CLI
  /go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_cli_suite_test.go:236
    MySQL application CSI via CLI
    /go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_cli_suite_test.go:243
  > Enter [BeforeAll] Backup and restore tests via OADP CLI - /go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_cli_suite_test.go:198 @ 07/25/25 19:03:46.863
2025/07/25 19:03:46 Verifying OADP CLI is available...
  [SKIPPED] OADP CLI not available: exit status 1, output: error: unknown command "oadp" for "kubectl"
  In [BeforeAll] at: /go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_cli_suite_test.go:204 @ 07/25/25 19:03:46.894
  < Exit [BeforeAll] Backup and restore tests via OADP CLI - /go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_cli_suite_test.go:198 @ 07/25/25 19:03:46.894 (31ms)
  > Enter [AfterEach] Backup and restore tests via OADP CLI - /go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_cli_suite_test.go:209 @ 07/25/25 19:03:46.894
2025/07/25 19:03:46 Post backup and restore state (CLI):  skipped
  [FAILED] Unexpected error:
      <*errors.errorString | 0xc0009520c0>: 
      resource name may not be empty
      {
          s: "resource name may not be empty",
      }
  occurred
  In [AfterEach] at: /go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_cli_suite_test.go:183 @ 07/25/25 19:03:46.925
  < Exit [AfterEach] Backup and restore tests via OADP CLI - /go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_cli_suite_test.go:209 @ 07/25/25 19:03:46.925 (31ms)
  > Enter [AfterAll] Backup and restore tests via OADP CLI - /go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_cli_suite_test.go:213 @ 07/25/25 19:03:46.925
2025/07/25 19:03:46 Checking if DPA is reconciled
2025/07/25 19:03:51 DPA status is Reconciled: True, reason Complete: Reconcile complete
2025/07/25 19:03:51 Waiting for Velero Pod to be running
2025/07/25 19:03:52 velero Pod phase is Pending
2025/07/25 19:03:57 velero Pod phase is Pending
2025/07/25 19:04:02 velero Pod phase is Pending
2025/07/25 19:04:07 velero Pod phase is Running
2025/07/25 19:04:07 Waiting for Node Agent pods to be running
2025/07/25 19:04:07 Checking for correct number of running Node Agent Pods...
2025/07/25 19:04:07 Checking if BSL is available
2025/07/25 19:04:07 BSL ts-velero-test-1 phase is Available
2025/07/25 19:04:07 Creating real DataProtectionTest before must-gather
2025/07/25 19:04:17 ✅ DPT e2e-uploadtest-dpt-1753470247 completed with phase: Complete
2025/07/25 19:04:17 skipMustGather: false
2025/07/25 19:04:17 Running OADP must-gather
  < Exit [AfterAll] Backup and restore tests via OADP CLI - /go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_cli_suite_test.go:213 @ 07/25/25 19:04:20.041 (33.116s)
S [SKIPPED] [33.177 seconds]
Backup and restore tests via OADP CLI [BeforeAll] [cli]
/go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_cli_suite_test.go:198
  Backup and restore applications via OADP CLI
  /go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_cli_suite_test.go:236
    MySQL application CSI via CLI
    /go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_cli_suite_test.go:243
```

false positive test not yet factually passing.

Joeavaikath committed:
Manual oadp-cli build and mv

Joeavaikath committed:
Fail cli e2e if oadp not found

kaovilai reviewed (commented):
lgtm

Joeavaikath commented:
/retest

Joeavaikath commented:
/retest

shubham-pampattiwar commented:
/override ci/prow/4.19-e2e-test-hcp-aws

shubham-pampattiwar commented:
/retest

Joeavaikath commented:
/retest

weshayutin commented:
@Joeavaikath DNM as is.. 
Looking at https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_oadp-operator/1835/pull-ci-openshift-oadp-operator-oadp-dev-4.19-e2e-test-cli-aws/1950561018925027328/artifacts/e2e-test-cli-aws/e2e/build-log.txt

Requirement would be that in the logs, you echo out the backup command used.   ATM we have
```
2025/07/30 15:53:52 Creating backup mysql-csi-cli-e2e-45b8909a-6d5d-11f0-b7c4-0a580a82f015 for case mysql-csi-cli-e2e via CLI
2025/07/30 15:53:52 Backup created via CLI: mysql-csi-cli-e2e-45b8909a-6d5d-11f0-b7c4-0a580a82f015
2025/07/30 15:54:02 backup phase: WaitingForPluginOperations
2025/07/30 15:54:12 backup phase: WaitingForPluginOperations
```

What I would NEED to see is:
```
Executing: oc oadp backup create -n mysql -blah -blah
2025/07/30 15:53:52 Creating backup mysql-csi-cli-e2e-45b8909a-6d5d-11f0-b7c4-0a580a82f015 for case mysql-csi-cli-e2e via CLI
2025/07/30 15:53:52 Backup created via CLI: mysql-csi-cli-e2e-45b8909a-6d5d-11f0-b7c4-0a580a82f015
2025/07/30 15:54:02 backup phase: WaitingForPluginOperations
2025/07/30 15:54:12 backup phase: WaitingForPluginOperations
```

kaovilai commented:
/lgtm

kaovilai commented:
/unlgtm

Joeavaikath committed:
Add logging before cli commands execute

Joeavaikath committed:
fix

kaovilai commented on tests/e2e/lib/cli_common.go:
```diff
@@ -0,0 +1,58 @@
+}
+
+func (c *CLICommand) LogCLICommand() {
+    args := []string{"kubectl", "oadp", c.Resource, c.Action}
```
I think wes want to see `kubectl oadp` or `oc oadp` after `:`

weshayutin replied:
I see:
```
2025/07/30 23:03:24 Executing CLI command: kubectl oadp backup create mysql-restic-cli-e2e-476f6f64-6d99-11f0-8da8-0a580a80f8c9 --include-namespaces mysql-persistent --default-volumes-to-fs-backup
```
this is good :)

weshayutin commented:
additionally, @kaovilai I think we can override the hcp failure here.  I need to follow up on the job.
error is weird:
```
2025/07/30 23:13:01 HC deletion timed out in WaitForHCDeletion, attempting to nuke resources with finalizers
2025/07/30 23:13:01 	Nuking HostedCluster
2025/07/30 23:13:01 	NUKE: Checking HostedCluster test-hc-agent in namespace clusters for finalizers
2025/07/30 23:13:01 	NUKE: Removing finalizers from HostedCluster test-hc-agent
2025/07/30 23:13:01 	NUKE: Removing finalizers from HostedControlPlane test-hc-agent
2025/07/30 23:13:01 	NUKE: Removing finalizers from Cluster test-hc-agent-test
2025/07/30 23:13:01 	NUKE: Removing finalizers from AgentCluster test-hc-agent
  [FAILED] failed to remove HCP: failed waiting for HostedCluster deletion: failed to wait for HC deletion (timeout: context deadline exceeded) and failed to nuke resources: 	NUKE: Error removing finalizers from AgentCluster test-hc-agent: Operation cannot be fulfilled on agentclusters.capi-provider.agent-install.openshift.io "test-hc-agent": the object has been modified; please apply your changes to the latest version and try again
  Unexpected error:
      <*errors.errorString | 0xc0004823a0>: 
      failed waiting for HostedCluster deletion: failed to wait for HC deletion (timeout: context deadline exceeded) and failed to nuke resources: 	NUKE: Error removing finalizers from AgentCluster test-hc-agent: Operation cannot be fulfilled on agentclusters.capi-provider.agent-install.openshift.io "test-hc-agent": the object has been modified; please apply your changes to the latest version and try again
      {
          s: "failed waiting for HostedCluster deletion: failed to wait for HC deletion (timeout: context deadline exceeded) and failed to nuke resources: \tNUKE: Error removing finalizers from AgentCluster test-hc-agent: Operation cannot be fulfilled on agentclusters.capi-provider.agent-install.openshift.io \"test-hc-agent\": the object has been modified; please apply your changes to the latest version and try again",
      }
  occurred
  In [It] at: /go/src/github.com/openshift/oadp-operator/tests/e2e/hcp_backup_restore_suite_test.go:56 @ 07/30/25 23:13:01.565
  < Exit [It] Agent HostedCluster backup and restore - /go/src/github.com/openshift/oadp-operator/tests/e2e/hcp_backup_restore_suite_test.go:144 @ 07/30/25 23:13:01.565 (14m16.118s)
  > Enter [AfterEach] HCP Backup and Restore tests - /go/src/github.com/openshift/oadp-operator/tests/e2e/hcp_backup_restore_suite_test.go:117 @ 07/30/25 23:13:01.565
2025/07/30 23:13:01 Deleting HostedCluster test-hc-agent in namespace clusters
2025/07/30 23:13:01 	Checking if HC test-hc-agent is deleted...
2025/07/30 23:13:01 	HC test-hc-agent is confirmed deleted
2025/07/30 23:13:01 Deleting HCP namespace clusters-test-hc-agent
2025/07/30 23:13:01 Deleting HostedControlPlane test-hc-agent in namespace clusters-test-hc-agent
2025/07/30 23:13:01 	Checking if HCP test-hc-agent is deleted...
2025/07/30 23:13:01 	HCP test-hc-agent still exists
2025/07/30 23:13:11 	Checking if HCP test-hc-agent is deleted...
2025/07/30 23:13:11 	HCP test-hc-agent is confirmed deleted
2025/07/30 23:13:11 	Checking if HCP test-hc-agent is deleted...
2025/07/30 23:13:11 	HCP test-hc-agent is confirmed deleted
2025/07/30 23:13:11 	HCP deleted
2025/07/30 23:13:11 Deleting secrets in namespace clusters-test-hc-agent
2025/07/30 23:13:11 	Waiting for the HC to be deleted
2025/07/30 23:13:11 	Attempting to verify HC deletion...
2025/07/30 23:13:11 	Checking if HC test-hc-agent is deleted...
2025/07/30 23:13:11 	HC test-hc-agent is confirmed deleted
2025/07/30 23:13:11 	HC deletion check result: true
2025/07/30 23:13:11 Post backup and restore state:  failed
```

Joeavaikath committed:
Tabs not spaces

kaovilai commented:
/lgtm
